pkg/fieldpath/fieldpath.go

```diff
@@ -57,3 +70,29 @@ func ExtractFieldPathAsString(obj interface{}, fieldPath string) (string, error)
 
 	return "", fmt.Errorf("unsupported fieldPath: %v", fieldPath)
 }
+
+// SplitMaybeSubscriptedPath checks whether the specified fieldPath is
+// subscripted, and
+//  - if yes, this function splits the fieldPath into path and subscript, and
+//    returns (path, subscript).
+//  - if no, this function returns (fieldPath, "").
+//
+// Example inputs and outputs:
+//  - "metadata.annotations['myKey']" --> ("metadata.annotations", "myKey")
+//  - "metadata.annotations['a[b]c']" --> ("metadata.annotations", "a[b]c")
+//  - "metadata.labels"               --> ("metadata.labels", "")
+//  - "metadata.labels['']"           --> ("metadata.labels", "")
```

Comment:
why single quotes vs double quotes vs no quotes? annotations and labels cannot have keys with `'`, `"`, `[`, `]`, right? would we eventually want the same syntax for field selectors?

Reply:
Actually, I don't think there's difference among them technically - no quotes should also work. The design has nothing to do with any changes on field selectors. Do you think using no quotes is more aligned with the current syntax?

There was some discussion on https://github.com/kubernetes/kubernetes/pull/36058#pullrequestreview-9308253.

Reply:
> The design has nothing to do with any changes on field selectors

see comment at https://github.com/kubernetes/kubernetes/pull/55902#discussion_r152098713 ... the implementation affects the field selection path

> Do you think using no quotes is more aligned with the current syntax?

All current field selectors are just `<field>[.<field>...]` paths... no index or key subscripts are currently used. Since this affects field selection, I wasn't sure what style we wanted for key subscripts. Was the single-quote syntax following jsonpath spec, etc?

Reply:
> see comment at #55902 (comment) ... the implementation affects the field selection path

I see. I couldn't find any docs on what fields are supported in `--field-selector` though. Yeah, we should probably support the key subscripts on these two fields for all objects. Does it make sense to do this separately and limit the scope of this PR to Downward API?

> All current field selectors are just <field>[.<field>...] paths... no index or key subscripts are currently used. Since this affects field selection, I wasn't sure what style we wanted for key subscripts. Was the single-quote syntax following jsonpath spec, etc?

Yes, single-quote follows the jsonpath spec.

Reply:
I argued about single-quotes too.  If I recall, it was Clayton who talked me down.

Reply:
According to http://goessner.net/articles/JsonPath/index.html#e2, dot-notation and bracket-notation (with single-quotes) are supported. `metadata.annotations['key']` uses both of them.

However, if we allow the dot-notation (e.g., `metadata.annotations.key`), we would have to escape the dots in the key, so I don't feel bad about the current approach.

Reply:
I guess a subset of jsonpath doesn't seem so bad... just looked odd at first

Reply:
> Yeah, we should probably support the key subscripts on these two fields for all objects. Does it make sense to do this separately and limit the scope of this PR to Downward API?

I'm not sure whether the intent of downward API was to always be aligned with field selection on pods or not. I'm not opposed to doing this prior to general field selection on labels/annotations, but we'd need to make sure we weren't modifying the API's field selection behavior as this PR does

Reply:
> we'd need to make sure we weren't modifying the API's field selection behavior as this PR does

The `AddFieldLabelConversionFunc` is used by both field selector and Downward API so I'm not sure how to avoid this side effect. The consequence of this PR is that `--field-selector=metadata.annotations['foo']=bar` becomes allowed but not working, while it's currently just disallowed. But the current behavior of field selection will not be affected.

I can create an issue to tackle the field selector issue after 1.9 and limit the scope of this PR to Downward API.

Reply:
> The consequence of this PR is that --field-selector=metadata.annotations['foo']=bar becomes allowed but not working, while it's currently just disallowed. But the current behavior of field selection will not be affected.

The server accepting a field selector it does not actually understand, and returning empty results for isn't something we can ship

Reply:
> The server accepting a field selector it does not actually understand, and returning empty results for isn't something we can ship

This is the list of fields allowed in field conversion for pod:
https://github.com/kubernetes/kubernetes/blob/5337ff8009d02fad613440912e540bb41e3a88b1/pkg/apis/core/v1/conversion.go#L160-L175

This is the list of fields allowed in field selector for pod:
https://github.com/kubernetes/kubernetes/blob/5337ff8009d02fad613440912e540bb41e3a88b1/pkg/registry/core/pod/strategy.go#L240-L246

If you look at the difference, `spec.serviceAccountName` and `status.hostIP` already have the same issue. I think we should figure out the field path in Downward API vs selector issue and fix it in a separate PR, instead of blocking this one.

Reply:
@liggitt, I asked @vishh to apply the `status/approved-for-milestone` label for now, but I'm happy to continue discussing what to do with this PR before LGTM. :)

Reply:
Having existing bugs does not warrant adding new ones.

I don't know the field selection code, but I would be surprised if we can easily add selection based on arbitrary annotation.  I am happy to be wrong.  It seems to me more likely that these two features, while sharing syntax, should NOT be conflated and should NOT use a single list of allowable strings.

Reply:
I'm looking into how to decouple the field selector and Downward API wrt the field conversion function, and will try again.

Reply:
> If you look at the difference, spec.serviceAccountName and status.hostIP already have the same issue

My guess is that those two resulted from additions for downward API as well. 

> I'm looking into how to decouple the field selector and Downward API wrt the field conversion function

the downward api expressions used the conversions so that pod specs in versions other than v1 would get converted to the proper field path. Decoupling from conversion might happen to work because we only have a single version of pod specs, but sets us up for conversion failures in the future. If we end up decoupling from conversion anyway, it should be conscious of the issue being introduced.

Reply:
I've added a separate conversion function for the field path used in Downward API. Please take a look at the second commit. We'd really like to have this PR in 1.9.

Reply:
thanks. looking at how this is used, all the places that call this want to know whether the expression was subscripted or not. they do that currently by checking the length of the returned subscript, which isn't really correct.

can we remove the ambiguity from this method by making it return `(path string, subscript string, hasSubscript bool)`, and make the callers treat the subscripted path distinctly? that avoids plumbing workarounds into API validation (allowing `metadata.annotations[]` and returning errors that reference that, etc) and depending on coincidences like the downward API allowing you to specify the entire `metadata.annotations` map in order to validate a subscripted path of `metadata.annotations['foo']`?